trace-dispatcher/doc/trace-dispatcher.md

```diff
@@ -0,0 +1,798 @@
+
+### Privacy
+
+__Privacy__ annotations allows limiting __trace-outs__ that particular traces can reach.  It is expressed in terms of:
+
+```haskell
+data Privacy
+    = Confidential | Public
+```
+
+__Confidential__ privacy level means that the trace will not be externalised from the system, except via __standard output__.
+
+The annotation mechanism is similar to the one of severity:
+
+```haskell
+privately :: Trace m a -> Trace m a
+withPrivacy :: Monad m => (a -> Privacy) -> Trace m a -> Trace m a
+```
+
+`Public` is the default __privacy__, in the absence of privacy annotations.
+
+Trace privacy cannot be configured.
+
+See [Confidentiality and privacy filtering implementation](#Confidentiality-and-privacy -filtering-implementation) for a more full discussion of semantics.
+
+To further prevent occasional leaks of `Confidential` traces, all output from those traces is tagged with the `CONFIDENTIAL` keyword.
+
+### Frequency
+
+__Frequency filtering__ is yet another part of __trace filtering__, and represents an optional limit on the observable frequency of individual trace messages.
+
+Semantically, this is corresponds to a randomly-fair suppression of messages within a particular trace, when their moving-average frequency exceeds a given threshold parameter.
+
+The __frequency limiter__ itself emits a __suppression summary__ message under the following conditions:
+
+* when it message suppression begins, and
+* when message suppression stops -- adding the number of suppressed messages.
+
+__Frequency limiters__ are given a name to identify its activity.
+
+```haskell
+limitFrequency
+  :: MonadIO m
+  => Double                  -- ^ messages per second
+  -> Text                    -- ^ name of this limiter
+  -> Trace m a               -- ^ the limited trace
+  -> Trace m LimitingMessage -- ^ a trace emitting the messages of the limiter
+  -> m (Trace m a)           -- ^ the original trace
+
+data LimitingMessage =
+    StartLimiting Text
+  | StopLimiting  Text Int
+```
+
+The frequency filtering is intended to be applied to a subset of traces (those known to be noisy).  For this subset of traces the frequency limit can be configured.
+
+## Presentation
+### Formatting
+
+The `LogFormatting` typeclass is used to describe __trace presentation__ -- mapping __traces__ to __metrics__ and __messages__.
+
+* The `forMachine` method is used for a machine readable representation, which can varied through detail level.
+  It requires an implementation to be provided by the trace author.
+
+* the `forHuman` method shall represent the message in human readable form.
+  It's default implementation defers to `forMachine`.
+
+* the `asMetrics` method shall represent the message as `0` to `n` metrics.
+  It's default implementation assumes no metrics. If a text is given it is
+  appended as last element to the namespace.
```

Comment:
I find the last sentence confusing. Perhaps something like:

```
Each metric can optionally specify a name space as a `Text`.
```

Though I wonder if there is a `type NameSapace = Text` somewhere :thinking:

Reply:
Yes: `type Namespace = [Text]`